internal/manifests/gateway.go

```diff
@@ -0,0 +1,314 @@
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   GatewayName(opts.Name),
+			Labels: l,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: pointer.Int32Ptr(1),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels.Merge(l, GossipLabels()),
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        GatewayName(opts.Name),
+					Labels:      labels.Merge(l, GossipLabels()),
+					Annotations: a,
+				},
+				Spec: podSpec,
+			},
+			Strategy: appsv1.DeploymentStrategy{
+				Type: appsv1.RollingUpdateDeploymentStrategyType,
+			},
+		},
+	}
+}
+
+// NewGatewayHTTPService creates a k8s service for the lokistack-gateway HTTP endpoint
+func NewGatewayHTTPService(opts Options) *corev1.Service {
+	serviceName := serviceNameGatewayHTTP(opts.Name)
+	l := ComponentLabels(LabelGatewayComponent, opts.Name)
+	a := serviceAnnotations(serviceName, opts.Flags.EnableCertificateSigningService)
+
+	return &corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: corev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        serviceName,
+			Labels:      l,
+			Annotations: a,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name: "metrics",
+					Port: httpPort,
+				},
+			},
+			Selector: l,
+		},
+	}
+}
+
+// gatewayConfigMap creates a configMap for rbac.yaml and tenants.yaml
+func gatewayConfigMap(opt Options) (*corev1.ConfigMap, string, error) {
+	cfg := gatewayConfigOptions(opt)
+	rbacConfig, tenantsConfig, err := gateway.Build(cfg)
+	if err != nil {
+		return nil, "", err
+	}
+
+	s := sha1.New()
+	_, err = s.Write(rbacConfig)
+	if err != nil {
+		return nil, "", err
+	}
+	sha1C := fmt.Sprintf("%x", s.Sum(nil))
+
+	return &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: corev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   LabelGatewayComponent,
+			Labels: commonLabels(opt.Name),
+		},
+		BinaryData: map[string][]byte{
+			gateway.LokiGatewayRbacFileName:   rbacConfig,
+			gateway.LokiGatewayTenantFileName: tenantsConfig,
+		},
+	}, sha1C, nil
+}
+
+// gatewayConfigOptions converts Options to gateway.Options
+func gatewayConfigOptions(opt Options) gateway.Options {
+	return gateway.Options{
+		Stack:     opt.Stack,
+		Namespace: opt.Namespace,
+		Name:      opt.Name,
+	}
+}
+
+func configureGatewayPKI(podSpec *corev1.PodSpec) error {
+	secretVolumeSpec := corev1.PodSpec{
+		Volumes: []corev1.Volume{
+			{
+				Name: "tls-secret",
+				VolumeSource: corev1.VolumeSource{
+					Secret: &corev1.SecretVolumeSource{
+						SecretName: LabelGatewayComponent,
+					},
+				},
+			},
+			{
+				Name: "tls-configmap",
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: LabelGatewayComponent,
+						},
+					},
+				},
+			},
+		},
```

Comment:
Who is actually creating this secret and this configmap? Is this PKI configuration dedicated for the ServiceMonitors to scrape through TLS or is this dedicated to access the Gateway via HTTPS? Former gets a certificate from the cert signining service, latter is something we need to figure out and leave some sort of third-party configurability.

Reply:
It's the 2nd case here (this is dedicated to access the Gateway via HTTPS). We need to figure out 3rd party configurability to inject tls `cert, key and ca`

Reply:
Alright, to summarize things going forward you still need to configure the `--web.internal.listen` with a PKI from the cert-signing-service to expose metrics via a TLS Service Monitor endpoint when the `--with-tls-service-monitor` flag is enabled. AFAICS, the observatorium-api does not provide `--tls.internal.*` flags to configure that PKI, so you need to make an upstream contribution for that.